### PR TITLE
Do not error when printing view created in no-grad modified in-place in no-grad

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113849
* __->__ #113716

Fixes https://github.com/pytorch/pytorch/issues/99968

### CODE REVIEW

File: torch/_tensor_str.py [hunk 1]
```diff
@@ -599,11 +599,22 @@ def indented_str(s, indent):
...
+    except RuntimeError:
+        # Accessing the grad_fn calls rebasing logic which would cause an error
+        # if that tensor is a view created in no-grad mode modified in-place in
+        # no-grad mode. See: https://github.com/pytorch/pytorch/issues/99968
+        grad_fn_name = "Undefined"
+
+    if grad_fn_name is None and grad_fn is not None:
+        grad_fn_name = type(grad_fn).__name__
+        if grad_fn_name == "CppFunction":
+            grad_fn_name = grad_fn.name().rsplit("::", 1)[-1]
```
Comment (albanD) [inline torch/_tensor_str.py:614]: You can leave this in the try so that you don't need the condition?

File: torch/_tensor_str.py [hunk 2]
```diff
@@ -599,11 +599,22 @@ def indented_str(s, indent):
...
-            name = inp.grad_fn.name().rsplit("::", 1)[-1]
-        suffixes.append(f"grad_fn=<{name}>")
+    grad_fn_name = None
+    try:
+        grad_fn = inp.grad_fn
+    except RuntimeError:
+        # Accessing the grad_fn calls rebasing logic which would cause an error
+        # if that tensor is a view created in no-grad mode modified in-place in
+        # no-grad mode. See: https://github.com/pytorch/pytorch/issues/99968
+        grad_fn_name = "Undefined"
```
Comment (albanD) [inline torch/_tensor_str.py:609]: It is not really undefined, it's "Invalid" ? Like anything trying to use it is going to crash.

Comment (soulitzer) [reply torch/_tensor_str.py:609]: Yeah "Invalid" does sound better

File: torch/_tensor_str.py (hunk 1 above)
Comment (soulitzer) [reply torch/_tensor_str.py:614]: If grad_fn.name() fails or doesn't return a string, wouldn't we rather have a legit error? Also, rsplit is a no-op anyway if we have a custom autograd Function, so we didn't need the condition in the first place, though maybe I prefer this being explicit.

File: torch/_tensor_str.py (hunk 2 above)
Comment (soulitzer) [reply torch/_tensor_str.py]: Updated, thanks!

File: torch/_tensor_str.py (hunk 1 above)
Comment (albanD) [reply torch/_tensor_str.py:614]: Ok.

**Review by albanD — APPROVED** [review]
SGTM

**Pull request closed without merging**